server/executor/queue.go

```diff
@@ -0,0 +1,474 @@
+			return nil
+		}
+
+		run, err := q.runs.GetRun(ctx, stopRequest.TestID, stopRequest.RunID)
+		if err != nil {
+			return fmt.Errorf("failed to get run %d for test %s: %w", stopRequest.RunID, stopRequest.TestID, err)
+		}
+
+		if run.State == test.RunStateStopped {
+			return nil
+		}
+
+		return q.cancelRunHandlerFn(ctx, run)
+
+	})
+
+	q.subscriptor.Subscribe((StopRequest{job.Test.ID, job.Run.ID}).ResourceID(), sfn)
+}
+
+func (q Queue) resolveTransaction(ctx context.Context, job Job) transaction.Transaction {
+	if q.transactions == nil {
+		return transaction.Transaction{}
+	}
+
+	tran, err := q.transactions.GetAugmented(ctx, job.Transaction.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return transaction.Transaction{}
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return tran
+}
+func (q Queue) resolveTransactionRun(ctx context.Context, job Job) transaction.TransactionRun {
+	if q.transactionRuns == nil {
+		return transaction.TransactionRun{}
+	}
+
+	tranRun, err := q.transactionRuns.GetTransactionRun(ctx, job.Transaction.ID, job.TransactionRun.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return transaction.TransactionRun{}
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return tranRun
+}
+
+func (q Queue) resolveTest(ctx context.Context, job Job) test.Test {
+	if q.tests == nil {
+		return test.Test{}
+	}
+
+	t, err := q.tests.GetAugmented(ctx, job.Test.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return test.Test{}
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return t
+}
+
+func (q Queue) resolveTestRun(ctx context.Context, job Job) test.Run {
+	if q.runs == nil {
+		return test.Run{}
+	}
+
+	run, err := q.runs.GetRun(ctx, job.Test.ID, job.Run.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return test.Run{}
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return run
+}
+
+func (q Queue) resolvePollingProfile(ctx context.Context, job Job) pollingprofile.PollingProfile {
+	if q.pollingProfiles == nil {
+		return pollingprofile.PollingProfile{}
+	}
+
+	profile, err := q.pollingProfiles.Get(ctx, job.PollingProfile.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return pollingprofile.PollingProfile{}
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return profile
+}
+
+func (q Queue) resolveDataStore(ctx context.Context, job Job) datastore.DataStore {
+	if q.dataStores == nil {
+		return datastore.DataStore{}
+	}
+
+	ds, err := q.dataStores.Get(ctx, job.DataStore.ID)
+	if errors.Is(err, sql.ErrNoRows) {
+		return datastore.DataStore{}
+	}
+	if err != nil {
+		panic(err)
+	}
+
+	return ds
+}
+
+func NewInMemoryQueueDriver(name string) *InMemoryQueueDriver {
+	return &InMemoryQueueDriver{
+		queue: make(chan Job),
+		exit:  make(chan bool),
+		name:  name,
+	}
+}
+
+type InMemoryQueueDriver struct {
+	queue chan Job
+	exit  chan bool
+	q     *Queue
+	name  string
+}
+
+func (r *InMemoryQueueDriver) SetQueue(q *Queue) {
+	r.q = q
+}
+
+func (r InMemoryQueueDriver) Enqueue(job Job) {
+	r.queue <- job
+}
+
+func (r InMemoryQueueDriver) Start() {
+	for i := 0; i < 5; i++ {
```

Comment:
The number `5` here is the number of internal workers for this driver, isn't it? Can we move it to a constant? (and maybe later to a tracetest config?)

Reply:
I'll add it as a constant, but I don't think this will have a lot of use outside testing